# EXP 2.3    
### Breast_cancer classification


In [52]:
# 필요한 모듈 import 하기
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# 데이터 준비
breast_cancer = load_breast_cancer()         


breast_cancer_data = breast_cancer.data             #feature data 지정
                                                    #breast_cancer_data라는 변수에 breast_cancer의 데이터 입력

breast_cancer_label = breast_cancer.target          #label data 지정
                                                    #breast_cancer의 label 정보를 변수에 저장

print('"target name은 %s 입니다."' %breast_cancer.target_names)            #target names 출력  
print("\n")                                                       #breast_cancer의 종류 알 수 있다 


# train, test 데이터 분리 
# 모델 학습과 테스트용 문제지와 정답지 준비 
X_train, X_test, y_train, y_test = train_test_split (breast_cancer_data,      # 모델의 feature 
                                                    breast_cancer_label,      # 모델이 맞춰야 하는 정답 (0-9까지의 숫자)
                                                    test_size = 0.2,   # 전체 데이터 중 20%를 test데이터로 지정
                                                    random_state=32)    # 랜덤으로 train/test data 분리 


"target name은 ['malignant' 'benign'] 입니다."




In [53]:
# target name이 ['malignant' 'benign'] 로 되어 있어 어떤 특징으로 분류 되었는지 알기 어렵다.
#              ['음성','양성']
# 유방암 데이터 분류 특징/기준

breast_cancer.feature_names

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

위와 같은 여러 특징을 통하여 유방암을 2가지(양섬,음성)으로 분류하는 것을 알 수 있다.

---
### 1) Decision Tree 모델

In [54]:
from sklearn.tree import DecisionTreeClassifier

#(5) 모델 학습 - X_train, y_train 활용하여 분류기 모델 만들기 / 성능 체크
decision_tree = DecisionTreeClassifier(random_state = 32)
decision_tree.fit(X_train,y_train)      # 모델학습 - training dataset에 맞게 패턴 파악 및 예측/학습
y_pred= decision_tree.predict(X_test)

print(classification_report(y_test,y_pred))

#성능평가
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_pred)
accuracy

              precision    recall  f1-score   support

           0       0.84      0.84      0.84        44
           1       0.90      0.90      0.90        70

    accuracy                           0.88       114
   macro avg       0.87      0.87      0.87       114
weighted avg       0.88      0.88      0.88       114



0.8771929824561403

---
### 2) Random Forest 모델

In [55]:
from sklearn.ensemble import RandomForestClassifier


random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))


from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_pred)
accuracy


              precision    recall  f1-score   support

           0       0.91      0.93      0.92        44
           1       0.96      0.94      0.95        70

    accuracy                           0.94       114
   macro avg       0.93      0.94      0.94       114
weighted avg       0.94      0.94      0.94       114



0.9385964912280702

---
### 3) SVM 모델

In [56]:
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_pred)
accuracy


              precision    recall  f1-score   support

           0       0.91      0.73      0.81        44
           1       0.85      0.96      0.90        70

    accuracy                           0.87       114
   macro avg       0.88      0.84      0.85       114
weighted avg       0.87      0.87      0.86       114



0.868421052631579

---
### 4) SGD Classifier 모델

In [57]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       0.86      0.82      0.84        44
           1       0.89      0.91      0.90        70

    accuracy                           0.88       114
   macro avg       0.87      0.87      0.87       114
weighted avg       0.88      0.88      0.88       114



0.8771929824561403

---
### 5) Logistic Regression 모델

In [58]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
# 오류나서 추가함 
logistic_model = LogisticRegression(solver='lbfgs', max_iter=10000).fit(X_train, y_train)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)

accuracy


              precision    recall  f1-score   support

           0       0.89      0.89      0.89        44
           1       0.93      0.93      0.93        70

    accuracy                           0.91       114
   macro avg       0.91      0.91      0.91       114
weighted avg       0.91      0.91      0.91       114



0.9122807017543859

---
# 분석하기 
정답(label)의 분포가 40-114로 불균형하다고 판단하여 accuracy 말고 다른 모델로 선택이 필요하다.  
여러가지 특징을 통하여 traget class로 분류 되어야하므로, 유방암의 정확한 특징을 놓치지 않고 양성/음성을 판단하는 과정이 필요할 것으로 보인다. 
양성인데 음성이라고 판단되어 수술 시기를 놓치는 상황이 적어야한다. 
  
  따라서 **Recall 이 중요하다 !!**


- 5가지 모델의 Recall 비교 
  - Decision Tree   : 0.87  
  - Random Forest: 0.94
  - SVM : 0.84
  - SGD Classifier : 0.87
  - Logistic Regression : 0.91  
  
Recall의 비교를 통하여 **Random Forest** 모델이 0.94로 가장 높게 나왔다.  
 
 

---
## 루브릭 평가 요구사항  
  
### 1) 3가지 데이터셋의 구성이 합리적으로 진행되었는가?
-  .data와 .target을 통해 feature data와 label data를 지정하였으며, .feature_names과 DESCR을 통하여 데이터 분류의 특징이 되는 기준을 알아보았습니다. 
  
### 2) 3가지 데이터셋에 대해 각각 5가지 모델을 성공적으로 적용하였는가?
-  각 데이터셋마다 Decision Tree, Random Forest, SVM, SGD Classifier, Logistic Regression 모델을 학습하고 테스트를 수행하였습니다.  

### 3) 3가지 데이터셋에 대해 모델의 평가지표가 적절히 선택되었는가? 
-  5가지 모델을 수행 후, 분석하여 각 데이터셋에 필요한 평가지표를 선정하고 모델의 성능을 평가하였습니다.  

---  
---  
# 회고  
  
  처음 모델 평가를 하게 되면 정확도만 중요하다고 생각이 들었다. 하지만 오차행렬을 학습하고 참이 거짓으로, 거짓이 참으로 판단되는 잘못된 경우도 있다는 것을 알아차리게 되었으며 각 상황에 따라 중요하게 평가되는 평가지표가 나타남을 알 수 있었다.  
- 음성을 놓치지 말아야 하는 상황일때, 음성이 양성으로 판단되는 것을 막아야하므로 FP가 낮을 수록 좋고 따라서 정밀도(Precision)가 높은 모델을 선택하면 좋다.  

- 양성을 놓치지 말아야 하는 상황일때, 양성이 음성으로 판단되는 것을 막아야하므로 FN이 낮을 수록 좋고 따라서 재현율(Recall)이 높은 모델을 선택하면 좋다.  

각각 다른 기준을 가진 데이터셋을 여러가지 평가지표로 다양한 방향으로 판단할 수 있었다.